comp/languagedetection/client/client.go

```diff
@@ -0,0 +1,328 @@
+	}
+	deps.Lc.Append(fx.Hook{
+		OnStart: cl.start,
+		OnStop:  cl.stop,
+	})
+
+	return util.NewOptional[Component](cl)
+}
+
+// start starts streaming languages to the Cluster-Agent
+func (c *client) start(_ context.Context) error {
+	c.logger.Infof("Starting language detection client")
+	go c.run()
+	return nil
+}
+
+// stop stops the client
+func (c *client) stop(_ context.Context) error {
+	c.cancel()
+	return nil
+}
+
+// run starts processing events and starts streaming
+func (c *client) run() {
+	defer c.logger.Infof("Shutting down language detection client")
+	// workloadmeta can't be initialized in the constructor or provided as a dependency until workloadmeta is refactored as a component
+	if c.store == nil {
+		c.store = workloadmeta.GetGlobalStore() // TODO(components): should be replaced by components
+	}
+
+	eventCh := c.store.Subscribe(
+		subscriber,
+		workloadmeta.NormalPriority,
+		workloadmeta.NewFilter(
+			[]workloadmeta.Kind{
+				workloadmeta.KindKubernetesPod, // Subscribe to pod events to clean up the current batch when a pod is deleted
+				workloadmeta.KindProcess,       // Subscribe to process events to populate the current batch
+			},
+			workloadmeta.SourceAll,
+			workloadmeta.EventTypeAll,
+		),
+	)
+	defer c.store.Unsubscribe(eventCh)
+
+	metricTicker := time.NewTicker(runningMetricPeriod)
+	defer metricTicker.Stop()
+
+	retryProcessWithoutPodTicker := time.NewTicker(c.retryProcessWithoutPodPeriod)
+	defer retryProcessWithoutPodTicker.Stop()
+
+	go c.startStreaming()
+
+	for {
+		select {
+		case eventBundle := <-eventCh:
+			c.processEvent(eventBundle)
+		case <-retryProcessWithoutPodTicker.C:
+			c.retryProcessEventsWithoutPod()
+		case <-metricTicker.C:
+			c.telemetry.Running.Set(1)
+		case <-c.ctx.Done():
+			return
+		}
+	}
+}
+
+// processEvent processes events from workloadmeta
+func (c *client) processEvent(evBundle workloadmeta.EventBundle) {
+	close(evBundle.Ch)
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.logger.Tracef("Processing %d events", len(evBundle.Events))
+	for _, event := range evBundle.Events {
+		switch event.Entity.GetID().Kind {
+		case workloadmeta.KindProcess:
+			c.processProcessEvent(event, false)
+		case workloadmeta.KindKubernetesPod:
+			c.processPodEvent(event)
+		}
+
+	}
+}
+
+// startStreaming retrieves the language detection client (= the DCA Client) and periodically sends data to the Cluster-Agent
+func (c *client) startStreaming() {
+	freshUpdateTimer := time.NewTicker(c.newUpdatePeriod)
+	defer freshUpdateTimer.Stop()
+
+	periodicFlushTimer := time.NewTicker(c.periodicalFlushPeriod)
+	defer periodicFlushTimer.Stop()
+
+	if c.langDetectionCl == nil {
+		// TODO(components): The ClusterAgentClient should most likely be a component. Moreover it should provide a retry mechanism or at least, the duration before the next try.
+		// Since currently we never retry `GetClusterAgentClient` in other parts of the code, we choose to follow the same pattern.
```

Comment:
You have this issue because you're trying to `GetClusterAgentClient` when you don't need it right now.
What you need is to have in `c` a `detectionClientAccesor func() (clusteragent.LanguageDetectionClient, error)` that you can set in unit tests and only call it.

With this every call to `send()` will retry to get with the current retrier.

Reply:
It's the only the comment but the whole logic!

Reply:
I forgot to push the logic change, now it should be done